### ISSUE TITLE
REPL rendering of nested vs-collections is.. poor

### ISSUE THREAD

**Issue by justinpombrio** [brownplt/pyret-lang#718]
This program defines a custom printer:

```
import valueskeleton as VS

data Nothin:
  | nothin
    with:
    _output(self):
      VS.vs-collection("Table", [list:
          VS.vs-collection("Headers", [list: "header"].map(VS.vs-value)),
          VS.vs-collection("Row1", [list: 1].map(VS.vs-value)),
          VS.vs-collection("Row2", [list: 2].map(VS.vs-value)),
          VS.vs-collection("Row3", [list:
              VS.vs-collection("Col1", [list: VS.vs-value("a")]),
              VS.vs-collection("Col1", [list: VS.vs-value("b")]),
              VS.vs-collection("Col1", [list: VS.vs-value("c")])
            ].map(VS.vs-value))])
    end
end
```

*Printing* `nothin` is almost correct, and displays:

```[Table: [Headers: header], [Row1: 1], [Row2: 2], [Row3: vs-collection(Col1, [list: vs-value(a)]), vs-collection(Col1, [list: vs-value(b)]), vs-collection(Col1, [list: vs-value(c)])]]```

(though it should not show "vs-collection").

However, the rendering in the repl is scrambled:

```
> nothin

[Table: [Row3: "header", 1, 2], [Row2: vs-collection("Col1"[list: vs-value("a")])], [Row1: vs-collection("Col1"[list: vs-value("b")])], [Headers: vs-collection("Col1"[list: vs-value("c")])]]
```

It's hard to tell what's going on here, but apparently rendering isn't done by recursive descent?

**Comment by jpolitz** [comment]
I think it's:

1.  Backwards, so there's a missing `.reverse()` somewhere
2.  Missing an unwrap for the nested vs-collection

**Comment by blerner** [comment]
Aw hell.  There's always a reversal problem in renderValueSkeleton, or its equivalent in CPO, which has to do with the stack-of-stacks printing mechanism: I thought I'd solved them all already :(  As for the nested vs-collections, I'm not sure why that's not working.

**Comment by justinpombrio** [comment]
The output isn't merely reversed, though. Parents in the tree are ending
with *different numbers of children* than they started with, that they've
stolen from other parents. Unless the whole tree is constructed via indices
or something?

**Comment by blerner** [comment]
Sortakinda.  All the nested values get pushed onto a worklist in recursive-descent order, then get filled into the waiting skeleton.  But if things get reversed somewhere, you can get the wrong numbers of filled-in items.

**Comment by justinpombrio** [comment]
Oops, there's a bug in the test. I wrote a superfluous ".map(VS.vs-value)"
that caused the "vs-collection" printing.

Here's a clearer test case:

```import valueskeleton as VS

data Nothin:
  | nothin
    with:
    _output(self):
      VS.vs-collection("0", [list:
          VS.vs-collection("1", [list: 2].map(VS.vs-value)),
          VS.vs-collection("3", [list: 4].map(VS.vs-value)),
          VS.vs-collection("5", [list: 6].map(VS.vs-value)),
          VS.vs-collection("7", [list:
              VS.vs-collection("8", [list: VS.vs-value(9)]),
              VS.vs-collection("10", [list: VS.vs-value(11)]),
              VS.vs-collection("12", [list: VS.vs-value(13)])
            ])])
    end
end```

> print(nothin)

[0: [1: 2], [3: 4], [5: 6], [7: [8: 9], [10: 11], [12: 13]]]

[0: [7: [12: 2], [10: 4], [8: 6]], [5: 9], [3: 11], [1: 13]]


On 5 May 2016 at 16:24, Justin Pombrio <justinpombrio@cs.brown.edu> wrote:

> The output isn't merely reversed, though. Parents in the tree are ending
> with *different numbers of children* than they started with, that they've
> stolen from other parents. Unless the whole tree is constructed via indices
> or something?
>
> On 5 May 2016 at 16:13, Joe Politz <notifications@github.com> wrote:
>
>> I think it's:
>>
>>    1. Backwards, so there's a missing .reverse() somewhere
>>    2. Missing an unwrap for the nested vs-collection
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/brownplt/pyret-lang/issues/718#issuecomment-217264761>
>>
>
>

**Comment by blerner** [comment]
So this is definitely a bug in CPO, and not in Pyret, since `tostring(nothin)` produces the correct output, right?

I think the problem is here: https://github.com/brownplt/code.pyret.org/blob/horizon/src/web/js/output-ui.js#L1406, which deliberately seems to go backwards down the list.  I don't have this branch checked out right now, but try reversing that (and the arithmetic two lines down) and see if it fixes your problem?

**Comment by blerner** [comment]
This should now be fixed in commit 5cbb540578c428c79f2fb0b18580c0464e10069a.  Certainly, the example above works properly!
